Support for multiple cut areas

AlexReimann opened the pull request:
Allows specifying multiple cut areas to exclude multiple areas in the field of view.

Usage:
Specify multiple pairs via a parameters:

```xml
cut_angles: [[start_1, stop_1], [start_2, stop_2]]
```

redkite commented on rslidar_pointcloud/src/rawdata.cc:
```diff
@@ -861,20 +872,32 @@ void RawData::unpack(const rslidar_msgs::rslidarPacket& pkt, pcl::PointCloud<pcl
 
-        if (start_clip_angle_< end_clip_angle_)
+        bool skip = false;
+        for (const auto& cliping : clip_angles_)
```
Typo in variable name.

Also, why don't you use structured bindings, if you change to C++17 anyway?

redkite commented on rslidar_pointcloud/src/rawdata.cc:
```diff
@@ -861,20 +872,32 @@ void RawData::unpack(const rslidar_msgs::rslidarPacket& pkt, pcl::PointCloud<pcl
         float arg_horiz_orginal = arg_horiz;
 
-        if (start_clip_angle_< end_clip_angle_)
+        bool skip = false;
```
```
if (clip_angles_.end() == find_if (clip_angles_.begin(), clip_angles_.end(), [&arg_horiz](const auto& p) {
  const auto& start = p.get<1>();
  const auto& end = p.get<2>();
  if (start < end) {
    if (arg_horiz < start || arg_horiz > end) {
      return false;
    }
  } else {
...
```

redkite reviewed (changes requested):
Might want to extract the nested loops into a function or lambda? For the sake of de-duplication maybe even rather an extra function?

redkite commented on rslidar_pointcloud/src/rawdata.cc:
```diff
@@ -88,19 +88,30 @@ void RawData::loadConfigFile(ros::NodeHandle node, ros::NodeHandle private_nh)
 
-  assert(std::abs(start_clip_angle_) <= 360.F);
-  assert(std::abs(end_clip_angle_) <= 360.F);
+  assert(start_clip_angles.size() == end_clip_angles.size());
```
That's useless. Runtime check only, since you don't have control over params, but the driver is always built with `Release` which should set `NDEBUG`

Commit:
[rslidar_pointcloud] Support for multiple clipping areas

AlexReimann commented:
Added parser to have an array of pair as input  Added runtime checks.
Changed back to c++11 just to not get problems with upstream.
Moved check into function

AlexReimann commented on rslidar_pointcloud/src/rawdata.cc:
```diff
@@ -88,19 +92,57 @@ void RawData::loadConfigFile(ros::NodeHandle node, ros::NodeHandle private_nh)
+    {
+      std::string fatal = "Parameter 'cut_angles' is has to be of type array";
+      ROS_FATAL_STREAM(fatal);
+      throw std::invalid_argument(fatal);
```
Sadly this is called during `on_init()` of the nodelet, which means it neither crashes the nodelet manager, nor does `ros::shutdown()` work.

redkite commented on rslidar_pointcloud/src/rawdata.cc:
```diff
@@ -88,19 +92,57 @@ void RawData::loadConfigFile(ros::NodeHandle node, ros::NodeHandle private_nh)
+  {
+    if (cut_pairs.getType() != XmlRpc::XmlRpcValue::TypeArray)
+    {
+      std::string fatal = "Parameter 'cut_angles' is has to be of type array";
```
Typo in message. Also should be const

AlexReimann replied:
Typo could be counted as sticking to the style of the file :D

redkite commented on rslidar_pointcloud/src/rawdata.cc:
```diff
@@ -88,19 +92,57 @@ void RawData::loadConfigFile(ros::NodeHandle node, ros::NodeHandle private_nh)
+    {
+       if (cut_pairs[i].getType() != XmlRpc::XmlRpcValue::TypeArray)
+       {
+         std::string fatal = "Parameter 'cut_angles' is has to be an array of arrays. Element with index "
```
const. Also has a typo

redkite commented on rslidar_pointcloud/src/rawdata.cc:
```diff
@@ -88,19 +92,57 @@ void RawData::loadConfigFile(ros::NodeHandle node, ros::NodeHandle private_nh)
+      if (cut_pairs[i][0].getType() != XmlRpc::XmlRpcValue::TypeDouble
+            || cut_pairs[i][1].getType() != XmlRpc::XmlRpcValue::TypeDouble)
+      {
+        std::string fatal = "Parameter 'cut_angles' is has to be an array of double pairs. Pair with index "
```
const

redkite commented on rslidar_pointcloud/src/rawdata.cc:
```diff
@@ -88,19 +92,57 @@ void RawData::loadConfigFile(ros::NodeHandle node, ros::NodeHandle private_nh)
+      ROS_INFO("Cut between horizontal angle range [%f, %f] degrees", start_cut_angle, end_cut_angle);
 
-  ROS_INFO("Clipping horizontal angle to range [%f, %f] degree", start_clip_angle_, end_clip_angle_);
+      start_cut_angle = static_cast<float>(start_cut_angle * M_PI / 180.0);
```
oO. static_cast to float and then write into double again? Why?

redkite replied:
Probably should cast in the `emplace_back` below?

redkite replied:
Or make a new variable that also includes the new format in its name.

redkite commented on rslidar_pointcloud/src/rawdata.cc:
```diff
@@ -1181,4 +1203,32 @@ void RawData::unpack_RS32(const rslidar_msgs::rslidarPacket& pkt, pcl::PointClou
   }
 }
 
+bool
```
We should stick to the formatting already used in this file

Roboterbastler commented:
@redkite @AlexReimann can we merge this? For bpearl we will have to rebase enway-devel onto newer master from upstream, so it would be easier/cleaner if this would be merged.

AlexReimann commented:
You can press the button, might change the commit ID on `enway-devel` and then needs another PR on the main repo?
Wasn't that smart to do it like this..

I belive @redkite is currently sick

AlexReimann commented:
Ahh, nope just adds another merge commit on top as it was "rebased" already
